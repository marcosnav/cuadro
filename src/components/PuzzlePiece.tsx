import React, { FC } from 'react';
import styled from 'styled-components';

interface IPiece {
  left: number;
  top: number;
}

interface IImage extends IPiece {
  image: string;
}

interface IProps {
  image: string;
  num: number;
  position: number;
}

const translate = (x: number, y: number) => `translate(${x}px,${y}px);`

const Piece = styled.div`
  background: cyan;
  flex-shrink: 0;
  height: 67px;
  left: 16px;
  overflow: hidden;
  position: absolute;
  top: 16px;
  transform: ${({top, left}: IPiece) => translate(left, top)};
  transition: all 0.2s ease;
  width: 67px;
`;

Piece.displayName = 'Piece';

const PieceImage = styled.div`
  background: gray;
  background-image: url(${({ image }: IImage) => image});
  background-position: center center;
  background-size: cover cover;
  height: 268px;
  position: relative;
  transform: ${({top, left}: IImage) => translate(left, top)};
  width: 268px;
`;

PieceImage.displayName = 'PieceImage';

/**
 * PuzzlePiece Component
 * {num} is the number in the board where it belongs
 * {position} is the current position of the piece in the board
 * @param props { num: number, position: number }
 */
const PuzzlePiece: FC<IProps> = ({ image, num, position }) => {
  const xAgent = (ref: number) => (ref % 4) * 67;
  const yAgent = (ref: number) => Math.floor(ref / 4) * 67;

  const pieceX = xAgent(position);
  const pieceY = yAgent(position);
  const imageX = -xAgent(num);
  const imageY = -yAgent(num);

  return (
    <Piece top={pieceY} left={pieceX}>
      <PieceImage image={image} top={imageY} left={imageX} />
    </Piece>
  );
};

export default PuzzlePiece;
