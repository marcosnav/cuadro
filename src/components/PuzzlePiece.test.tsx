import { mount } from 'enzyme';
import 'jest-enzyme';
import 'jest-styled-components';
import React from 'react';
import PuzzlePiece from './PuzzlePiece';

describe('<PuzzlePiece/>', () => {
  test('renders properly', () => {
    const component = mount(<PuzzlePiece image={''} num={1} position={3} />);
    expect(component).toMatchSnapshot();
  });

  test('must have a translation based on its position', () => {
    const component = mount(<PuzzlePiece image={''} num={1} position={3} />);
    expect(component).toHaveStyleRule('transform', 'translate(201px,0px)');

    component.setProps({ position: 6 });
    expect(component).toHaveStyleRule('transform', 'translate(134px,67px)');

    component.setProps({ position: 9 });
    expect(component).toHaveStyleRule('transform', 'translate(67px,134px)');

    component.setProps({ position: 12 });
    expect(component).toHaveStyleRule('transform', 'translate(0px,201px)');
  });

  test('must display the image based on the piece number', () => {
    const component = mount(<PuzzlePiece image={''} num={1} position={3} />);
    const image = component.find('PieceImage').first();
    expect(image).toHaveStyleRule('transform', 'translate(-67px,0px)');
  });
});
