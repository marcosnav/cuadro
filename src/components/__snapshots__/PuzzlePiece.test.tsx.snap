// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PuzzlePiece/> renders properly 1`] = `
.c0 {
  background: cyan;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  height: 67px;
  left: 16px;
  overflow: hidden;
  position: absolute;
  top: 16px;
  -webkit-transform: translate(201px,0px);
  -ms-transform: translate(201px,0px);
  transform: translate(201px,0px);
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  width: 67px;
}

.c1 {
  background: gray;
  background-image: url();
  background-position: center center;
  background-size: cover cover;
  height: 268px;
  position: relative;
  -webkit-transform: translate(-67px,0px);
  -ms-transform: translate(-67px,0px);
  transform: translate(-67px,0px);
  width: 268px;
}

<PuzzlePiece
  image=""
  num={1}
  position={3}
>
  <Piece
    left={201}
    top={0}
  >
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bdVaJa",
            "isStatic": false,
            "lastClassName": "c0",
            "rules": Array [
              "
  background: cyan;
  flex-shrink: 0;
  height: 67px;
  left: 16px;
  overflow: hidden;
  position: absolute;
  top: 16px;
  transform: ",
              [Function],
              ";
  transition: all 0.2s ease;
  width: 67px;
",
            ],
          },
          "displayName": "Piece",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bdVaJa",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
      left={201}
      top={0}
    >
      <div
        className="c0"
      >
        <PieceImage
          image=""
          left={-67}
          top={-0}
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-bwzfXH",
                  "isStatic": false,
                  "lastClassName": "c1",
                  "rules": Array [
                    "
  background: gray;
  background-image: url(",
                    [Function],
                    ");
  background-position: center center;
  background-size: cover cover;
  height: 268px;
  position: relative;
  transform: ",
                    [Function],
                    ";
  width: 268px;
",
                  ],
                },
                "displayName": "PieceImage",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-bwzfXH",
                "target": "div",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            image=""
            left={-67}
            top={-0}
          >
            <div
              className="c1"
            />
          </StyledComponent>
        </PieceImage>
      </div>
    </StyledComponent>
  </Piece>
</PuzzlePiece>
`;
